### PR TITLE
Raise 404 on unauthorized organization requests in the project scope.

### LINKED ISSUE #3807: Tokens are able to enumerate project names regardless of organization membership.
We respond with 404s when there are missing project names, but 403s when there is an existing project. Probably should just check for organization membership before even going into the project fetching.

Original HackerOne Report
====================
Title:	Project Enumeration of Other Teams
Types:	Information Disclosure
Link:	https://hackerone.com/reports/154048
Date:	2016-07-26 05:25:05 -0700
By:		vagg-a-bond

Details:
Hello,

It is possible for any valid user of @sentry to enumerate project names in other team by comparing errors. 

For example, if user user accesses ```GET /api/0/projects/sumne/project_sumne_001/``` which is not of his/her, then error will be ```403 FORBIDDEN``` and for  ```GET /api/0/projects/sumne/project_sumne_001_not_there/``` application responds with ```404 NOT FOUND```

So an attacker can enumerate all the projects available in a team by automating the process to request ```GET /api/0/projects/TARGET_TEAM_NAME/PROJECT_NAME/```. 

__Vulnerable Endpoint:__
> GET /api/0/projects/sumne/project_sumne_001/*

__Recommended Fix:__
Respond with ```404 NOT FOUND``` for both.

### PR DESCRIPTION
**Pull request by JTCunning**
Fixes #3807.

cc @getsentry/infrastructure.

---

This change is [Reviewable](https://reviewable.io/reviews/getsentry/sentry/3810)

### CODE REVIEW

**Comment by mattrobenolt** [comment]
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/sentry/api/bases/project.py, line 74 \[r1\]](https://reviewable.io:443/reviews/getsentry/sentry/3810#-KNcSoaGIu2Wm2NVio6T:-KNcSoaGIu2Wm2NVio6U:1112527862) ([raw file](https://github.com/getsentry/sentry/blob/2463a5c46ab2a5c9ac1e11564f701eca759b3fd3/src/sentry/api/bases/project.py#L74)):*
> ```Python
> 
>         org = Organization.objects.get_from_cache(slug=organization_slug)
>         if not access.from_request(request, org).memberships:
> ```

It seems more sane to me to test this before even querying the Project.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/sentry/3810)*

File: src/sentry/api/bases/project.py [hunk 1]
```diff
@@ -61,7 +61,7 @@ class ProjectEndpoint(Endpoint):
     def convert_args(self, request, organization_slug, project_slug, *args, **kwargs):
         try:
             org = Organization.objects.get_from_cache(slug=organization_slug)
-            if not access.from_request(request, org).memberships:
+            if request.user and not access.from_request(request, org).memberships:
```
Comment (mattrobenolt) [inline src/sentry/api/bases/project.py]: Does this mean API key auth won't follow this path?

Comment (JTCunning) [inline src/sentry/api/bases/project.py]: Hmm.. yea. Did this to fix up some tests but should probably make the actual method more resilient.

**Comment by JTCunning** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/sentry/api/bases/project.py, line 74 \[r1\]](https://reviewable.io:443/reviews/getsentry/sentry/3810#-KNcSoaGIu2Wm2NVio6T:-KNcrmhRB_-mjw8-7rbV:1964245312) ([raw file](https://github.com/getsentry/sentry/blob/2463a5c46ab2a5c9ac1e11564f701eca759b3fd3/src/sentry/api/bases/project.py#L74)):*
<details><summary><i>Previously, mattrobenolt (Matt Robenolt) wrote…</i></summary>
> It seems more sane to me to test this before even querying the Project.
</details>
Moved.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/sentry/3810)*

**Comment by mattrobenolt** [comment]
This seems correct. I assume we'd want this same behavior across other endpoints too. Just assuming they'd all be behaving the same way.

---

Reviewed 1 of 2 files at r1, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/getsentry/sentry/3810#-:-KNhDVoH4aQSvcNJ2xKZ:633310721)*

**Pull request merged**